src/materialized/lib.rs

```diff
@@ -196,20 +201,38 @@ pub fn serve(mut config: Config) -> Result<Server, failure::Error> {
     let switchboard = Switchboard::new(config.addresses, config.process, executor.clone());
 
     // Launch task to serve connections.
-    let mut mux = Mux::new();
-    mux.add_handler(switchboard.clone());
-    if is_primary {
-        // The primary is responsible for pgwire and HTTP traffic in
-        // addition to switchboard traffic.
-        let cmd_tx = Arc::downgrade(&cmd_tx);
-        let tls = match config.tls {
-            None => None,
-            Some(tls_config) => Some(tls_config.acceptor()?),
-        };
-        mux.add_handler(pgwire::Server::new(tls.clone(), cmd_tx.clone()));
-        mux.add_handler(http::Server::new(tls, cmd_tx, start_time));
-    }
-    runtime.spawn(async move { mux.serve(listener.incoming()).await });
+    //
+    // The lifetime of this task is controlled by two triggers that activate on
+    // drop. Draining marks the beginning of the server shutdown process and
+    // indicates that new user connections (i.e., pgwire and HTTP connections)
+    // should be rejected. Gracefully handling existing user connections
+    // requires that new system (i.e., switchboard) connections continue to be
+    // routed whiles draining. The shutdown trigger indicates that draining is
+    // complete, so switchboard traffic can cease and the task can exit.
+    let (drain_trigger, drain_tripwire) = Tripwire::new();
+    let (shutdown_trigger, shutdown_tripwire) = Tripwire::new();
+    runtime.spawn({
+        let switchboard = switchboard.clone();
+        async move {
+            let incoming = &mut listener.incoming();
+
+            // The primary is responsible for pgwire and HTTP traffic in
+            // addition to switchboard traffic until draining starts.
+            if is_primary {
+                let mut mux = Mux::new();
+                mux.add_handler(switchboard.clone());
+                mux.add_handler(pgwire::Server::new(tls.clone(), cmdq_tx.clone()));
+                mux.add_handler(http::Server::new(tls, cmdq_tx, start_time));
+                mux.serve(incoming.take_until(drain_tripwire)).await;
```

Comment:
this is beautiful.

Reply:
:D this was the fourth attempt to sort this out cleanly, so I'm glad you like it!